File: src/services/evm.js
import { Contract, JsonRpcProvider, parseEther, parseUnits } from 'ethers';
import { getEvmNetworkConfig } from '../config/evmNetworks.js';

const ERC20_ABI = [
  'function decimals() view returns (uint8)',
  'function symbol() view returns (string)',
  'function transfer(address to, uint256 amount) returns (bool)',
];

export function createEvmProvider(network, rpcOverride) {
  const { chainId, rpcUrl, name } = getEvmNetworkConfig(network, rpcOverride);
  const provider = new JsonRpcProvider(rpcUrl, chainId);
  return { provider, networkName: name, chainId, rpcUrl };
}

function maybeGasPrice(gwei) {
  if (!gwei) return undefined;
  return parseUnits(String(gwei), 'gwei');
}

function maybeGasLimit(limit) {
  if (!limit) return undefined;
  const n = BigInt(limit);
  if (n <= 0n) throw new Error('gasLimit must be > 0');
  return n;
}

export async function transferEth({ provider, signer, to, amount, gasPrice, gasLimit }) {
  const txReq = {
    to,
    value: parseEther(String(amount)),
  };
  const gp = maybeGasPrice(gasPrice);
  const gl = maybeGasLimit(gasLimit);
  if (gp) txReq.gasPrice = gp;
  if (gl) txReq.gasLimit = gl;

  const tx = await signer.sendTransaction(txReq);
  const receipt = await tx.wait();
  return { hash: tx.hash, receipt };
}

export async function transferErc20({ provider, signer, token, to, amount, decimals, gasPrice, gasLimit }) {
  const contract = new Contract(token, ERC20_ABI, signer);
  let tokenDecimals = decimals;
  if (tokenDecimals === undefined || tokenDecimals === null) {
    tokenDecimals = await contract.decimals();
  }
  const amt = parseUnits(String(amount), tokenDecimals);
  const overrides = {};
  const gp = maybeGasPrice(gasPrice);
  const gl = maybeGasLimit(gasLimit);
  if (gp) overrides.gasPrice = gp;
  if (gl) overrides.gasLimit = gl;

  const tx = await contract.transfer(to, amt, overrides);
  const receipt = await tx.wait();
  return { hash: tx.hash, receipt, decimals: tokenDecimals };
}

export async function getErc20Meta({ provider, token }) {
  const contract = new Contract(token, ERC20_ABI, provider);
  const [decimals, symbol] = await Promise.all([
    contract.decimals(),
    contract.symbol().catch(() => undefined),
  ]);
  return { decimals, symbol };
}

File: src/services/evm.rs
use std::sync::Arc;

use anyhow::{anyhow, Result};
use ethers::prelude::*;
use ethers::types::{Address, U256};
use ethers::utils::{parse_ether, parse_units};

use crate::config::evm_networks::{get_evm_network_config, EvmNetworkConfig};

abigen!(
    IERC20,
    r#"[
        function decimals() view returns (uint8)
        function symbol() view returns (string)
        function transfer(address to, uint256 amount) returns (bool)
    ]"#
);

pub struct Erc20Meta {
    pub decimals: u8,
    pub symbol: Option<String>,
}

pub fn create_evm_provider(
    network: &str,
    rpc_override: Option<&str>,
) -> Result<(Provider<Http>, EvmNetworkConfig)> {
    let cfg = get_evm_network_config(network, rpc_override)?;
    let provider = Provider::<Http>::try_from(cfg.rpc_url.as_str())
        .map_err(|err| anyhow!("Invalid RPC URL: {err}"))?;
    Ok((provider, cfg))
}

pub async fn transfer_eth(
    provider: Provider<Http>,
    wallet: LocalWallet,
    to: &str,
    amount: &str,
    gas_price: Option<&str>,
    gas_limit: Option<u64>,
    chain_id: u64,
) -> Result<TxHash> {
    let to_addr: Address = to.parse().map_err(|err| anyhow!("Invalid recipient address: {err}"))?;
    let value = parse_ether(amount)?;

    let wallet = wallet.with_chain_id(chain_id);
    let client = SignerMiddleware::new(provider, wallet);
    let client = Arc::new(client);

    let mut tx = TransactionRequest::new().to(to_addr).value(value);
    if let Some(gp) = maybe_gas_price(gas_price)? {
        tx = tx.gas_price(gp);
    }
    if let Some(gl) = gas_limit {
        tx = tx.gas(gl);
    }

    let pending = client.send_transaction(tx, None).await?;
    let tx_hash = pending.tx_hash();
    let _receipt = pending.await?;
    Ok(tx_hash)
}

pub async fn transfer_erc20(
    provider: Provider<Http>,
    wallet: LocalWallet,
    token: &str,
    to: &str,
    amount: &str,
    decimals: Option<u8>,
    gas_price: Option<&str>,
    gas_limit: Option<u64>,
    chain_id: u64,
) -> Result<TxHash> {
    let token_addr: Address = token.parse().map_err(|err| anyhow!("Invalid token address: {err}"))?;
    let to_addr: Address = to.parse().map_err(|err| anyhow!("Invalid recipient address: {err}"))?;

    let wallet = wallet.with_chain_id(chain_id);
    let client = SignerMiddleware::new(provider, wallet);
    let client = Arc::new(client);

    let contract = IERC20::new(token_addr, client.clone());
    let token_decimals = match decimals {
        Some(v) => v,
        None => contract.decimals().call().await?,
    };
    let amount_units = parse_units(amount, token_decimals as u32)?.into();

    let mut call = contract.transfer(to_addr, amount_units);
    if let Some(gp) = maybe_gas_price(gas_price)? {
        call = call.gas_price(gp);
    }
    if let Some(gl) = gas_limit {
        call = call.gas(gl);
    }

    let pending = call.send().await?;
    let tx_hash = pending.tx_hash();
    let _receipt = pending.await?;
    Ok(tx_hash)
}

pub async fn get_erc20_meta(
    provider: Provider<Http>,
    token: &str,
) -> Result<Erc20Meta> {
    let token_addr: Address = token.parse().map_err(|err| anyhow!("Invalid token address: {err}"))?;
    let contract = IERC20::new(token_addr, provider.into());
    let decimals = contract.decimals().call().await?;
    let symbol = contract.symbol().call().await.ok();
    Ok(Erc20Meta { decimals, symbol })
}

fn maybe_gas_price(gwei: Option<&str>) -> Result<Option<U256>> {
    let Some(value) = gwei else { return Ok(None) };
    let parsed = parse_units(value, "gwei")?.into();
    Ok(Some(parsed))
}
